Define @attributes on expressions (#450)
unicode-org/message-format-wg: spec/syntax.md

An **_<dfn>expression</dfn>_** is a part of a _message_ that will be determined
during the _message_'s formatting.

An _expression_ MUST begin with U+007B LEFT CURLY BRACKET `{`
and end with U+007D RIGHT CURLY BRACKET `}`.
An _expression_ MUST NOT be empty.

> An _expression_ can contain an _operand_,

Reviewer: I think we should promote "can" to normative "MUST" here. It was previously fine as "can" because there was nothing else permitted besides these three options.

```suggestion
An _expression_ MUST contain an _operand_,
```

Author: This is beyond the scope of this PR, which for this line is only removing the trailing space.

Author: Sure, that makes sense. I had not earlier caught the concern about an expression consisting of only attributes being non-empty.

```abnf
reserved       = reserved-start reserved-body
reserved-start = "!" / "#" / "%" / "*" / "<" / ">" / "/" / "?" / "~"

reserved-body  = *( [s] 1*(reserved-char / reserved-escape / quoted))
reserved-char  = %x00-08        ; omit HTAB and LF
               / %x0B-0C        ; omit CR
               / %x0E-19        ; omit SP
               / %x21-3F        ; omit @
               / %x41-5B        ; omit \
               / %x5D-7A        ; omit { | }
               / %x7E-D7FF      ; omit surrogates
               / %xE000-10FFFF
```

### Expression Attributes

An **_<dfn>expression attribute</dfn>_** is a key-value pair
containing a named argument that is attached to the _expression_ as a whole.

> They are intended to convey information about the _expression_
> that is not necessarily applicable as a _function_ _option_.

Reviewer: I think this is too opinionated for us to include at this juncture. I would tone it down:

```suggestion
These might be used to convey information about the _expression_ that are not necessarily applicable as _options_ to a _function_, such as attributes related handling during the translation process.
Or they might be used to override contextual or implementation-specific values used by a _function_ during formatting, such as the locale, time zone, or other settings.
```

For example, the `@locale` attribute overrides the locale of the _message_ for its _expression_.
Attributes like `@can-delete` and `@translate` might not have any impact on formatting,

> acting instead as instructions to translators.

Reviewer: The syntax specification should only include attributes that we formally define all the way through. If these are examples, they should be treated as such, e.g. "If an implementation supported attributes such as..."

_Expression attributes_ are prefixed by a U+0040 COMMERCIAL AT `@` sign,
but otherwise use the same syntax as _options_.

The _name_ of an _expression attribute_ MUST either be one of the following,

> or it MUST contain at least one U+002D HYPHEN-MINUS `-` character:

Reviewer: I think you're trying to create a reserved namespace of "no hyphens" for use by the specification and then allow implementations to define their own expression attributes by forcing them to include hyphens. If that's the case, _we should say that_. Personally, I would have gone for some other mechanism, such as reserving `_` (which is a valid `name` starter) for implementation defined items, e.g.:

> {$foo :function option=value @locale=fr-FR @canReorder=no @_custom=val}

Author: The reason for creating the namespace is explicitly said a few lines further down:
> To ensure future extensibility of this specification,
all other _expression attributes_ with a _name_ not containing `-` MUST be ignored.

An underscore `_` is a possible alternative as well. I went with `-` because that's the [requirement](https://html.spec.whatwg.org/multipage/custom-elements.html#valid-custom-element-name) used by custom HTML elements, and that seemed like pretty decent prior art to follow.

Reviewer: The starts-with-underscore I find unattractive, but it fits with `name` without any changes and only requires a one-character lookahead to figure it out. I like `-` okay, I guess, but we really should resolve our naming issues so that we don't have a motley collection of naming schemes.

I filed #453 to track.

> - `language`: SHOULD resolve to a sequence of IETF BCP 47 language tags;

Reviewer: `language` or `locale`??

Also, in formatting.md you said it MUST resolve to a language tag. It should say MUST here too. The syntax won't be responsible for checking it (in terms of ABNF) but let's not confuse people by mixing normative language.

Or avoid the issue by saying:

```suggestion
- `locale`: a sequence of well-formed IETF BCP47 language tags
```

  see _expression attribute resolution_.

> - `translate`: SHOULD have a literal value of either `yes` or `no`.

Reviewer: Same comment here.

Author: ```suggestion
- `translate`: a literal value of either `yes` or `no`.
```

To ensure future extensibility of this specification,

> all other _expression attributes_ with a _name_ not containing `-` MUST be ignored.

Reviewer: How about just allowing unrecognized attributes to be ignored?

Author: Because then we could not later assign a meaning to them without potentially breaking user code.

Reviewer: Reserving a namespace is a separate thing from ignoring unrecognized values. These can work in tandem. An MF2.0 implementation would ignore the unrecognized value `@foo` that MF2.1 unreserves, as well as `@_bar` that some other proprietary implementation is using.

Alternatively (and as I suggested elsewhere), as long as the values are syntactically correct, the implementation can just parse the values without imputing meaning to them and pass them. It only has to apply validation rules to attributes it "knows" (and do we define an error for that???)

Author: Consider the following hypothetical, in a world where we allow implementations to assign their own meaning to any attributes not explicitly defined in the spec like `@locale` and `@translate`:

Let's say that an implementation defines `@fallback` to provide a value that should be used for a placeholder, should its formatting fail for whatever reason. This doesn't really make sense as a function option, because it's covering the behaviour of what happens if the function throws an error. Other implementations could just ignore this attribute, and use the spec-defined fallback representation.

Then let's say that in a 2.1 version of the MF spec we'd like to define `@fallback` to mean a fallback locale to use, if the function doesn't support the message locale, or the one set by `@locale`. This hypothetical attribute would have the same semantics as `@locale`, but it would append values to the language priority list rather than replacing it.

But in this world we could not do that without potentially breaking userspace, because an implementation as defined above could exist, along with a pre-existing body of messages using `@fallback` with a completely different meaning. In this case, the meanings of "fallback" are quite different, but even should they align, it's very likely that the detailed implementation and requirements could differ for the attributes. In other words, we would not be able to define _any_ future attributes in the spec without a danger of breaking userspace.

Multiple _expression attributes_ are permitted in an _expression_.
Each _expression attribute_ is separated by whitespace.
If present, _expression attributes_ MUST be the last non-whitespace content in an _expression_,

> and MUST NOT be followed by any _function_ _options_.

Reviewer: Well... or anything else, like an annotation with no options or an operand. Just stop after "content in an _expression_"?

> If present, _expression attributes_ MUST be the last non-whitespace content in an _expression_,
> and MUST NOT be followed by any _function_ _options_.

Author: ```suggestion
If present, _expression attributes_ MUST be the last non-whitespace content in an _expression_.
```

```abnf
expression-attribute = "@" option
```

> Examples of _expressions_ with _expression attributes_
>
> A _message_ in English including a French _literal_ that should not be translated:
>
> ```
> {In French, "{|bonjour| @locale=fr @translate=no}" is a greeting}
> ```
>

> > A _message_ with a `$date` _variable_ formatted with the `:datetime` _function_
> > using the root "und" locale:
> >
> > ```
> > {Today is {$date :datetime weekday=long @locale=und}.}

Reviewer: The root locale might not be the most evocative? Perhaps mirror the above example:

```suggestion
> A _message_ with a `$date` _variable_ formatted using the `:datetime` _function_
> and overriding the locale to use French formatting:
>```
>{The date would be displayed in French as {$date :datetime @locale=fr}
```

(Note that this doesn't require "translate=no" because the datetime function is doing the work.
```

> ```
>
> A _message_ with a `$start` _variable_ that should in translations

> > be kept as the first element of the message.

Reviewer: should in translations be kept => should be kept in translations

> > A _message_ with a `$start` _variable_ that should in translations
> > be kept as the first element of the message.

Author: ```suggestion
> A _message_ with a `$start` _variable_ that should be kept in translations
> as the first element of the message.
```
